packages/ampd-sdk/src/event/event_handler.rs

```diff
@@ -0,0 +1,520 @@
+
+use async_trait::async_trait;
+use cosmrs::Any;
+use error_stack::{report, Context, Report, ResultExt};
+use events::{AbciEventTypeFilter, Event};
+use futures::{pin_mut, Stream, TryFutureExt};
+use mockall::automock;
+use report::ErrorExt;
+use serde::{Deserialize, Serialize};
+use thiserror::Error;
+use tokio::time::interval;
+use tokio_stream::{Elapsed, StreamExt};
+use tokio_util::sync::CancellationToken;
+use tracing::{error, info};
+use typed_builder::TypedBuilder;
+
+use crate::event::callbacks::{
+    default_broadcast_error_cb, default_event_subscription_error_cb, default_handler_error_cb,
+};
+use crate::future::{with_retry, RetryPolicy};
+use crate::grpc::client;
+
+#[automock(
+    type Err = Error;
+    type Event = Event;
+)]
+#[async_trait]
+pub trait EventHandler: Send + Sync {
+    type Err: Context;
+    type Event: TryFrom<Event>;
+
+    async fn handle(
+        &self,
+        event: &Self::Event,
+        token: &CancellationToken,
+    ) -> Result<Vec<Any>, Self::Err>;
+
+    fn subscription_params(&self) -> SubscriptionParams;
+}
+
+pub struct SubscriptionParams {
+    event_filters: Vec<AbciEventTypeFilter>,
+    include_block_begin_end: bool,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct Config {
+    #[serde(with = "humantime_serde")]
+    pub stream_timeout: Duration,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            stream_timeout: Duration::from_secs(10),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+#[allow(dead_code)]
+pub enum Error {
+    #[error("failed to retrieve events stream from the client")]
+    EventStream,
+
+    #[error("timeout while waiting for event stream")]
+    StreamTimeout(#[from] Elapsed),
+
+    #[error("unable to parse event of type")]
+    EventConversion,
+
+    #[error("error when handling event messages")]
+    HandlerFailed,
+
+    #[error("error when broadcasting message")]
+    BroadcastFailed,
+
+    #[error("{0} task aborted")]
+    TaskAborted(String),
+}
+
+#[allow(dead_code)]
+pub enum HandlerTaskAction {
+    Abort,
+    Continue,
+    ContinueAndSleep(Duration),
+}
+
+#[derive(TypedBuilder)]
+#[allow(dead_code)]
+pub struct HandlerTask<H>
+where
+    H: EventHandler,
+    <H::Event as TryFrom<Event>>::Error: Context,
+    H::Event: Display,
+{
+    handler: H,
+    config: Config,
+    #[builder(default = String::from("default_task"))]
+    task_label: String,
+    #[builder(default = RetryPolicy::NoRetry)]
+    handler_retry_policy: RetryPolicy,
+    #[builder(default = RetryPolicy::NoRetry)]
+    broadcast_retry_policy: RetryPolicy,
+    #[builder(default = default_event_subscription_error_cb)]
+    event_subscription_error_cb: fn(Report<Error>) -> HandlerTaskAction,
+    #[builder(default = default_handler_error_cb)]
+    handler_error_cb: fn(&H::Event, H::Err) -> HandlerTaskAction,
+    #[builder(default = default_broadcast_error_cb)]
+    broadcast_error_cb: fn(&[Any], H::Err) -> HandlerTaskAction,
```

Comment:
```suggestion
    broadcast_error_cb: fn(&Any, Report<Error>) -> HandlerTaskAction,
```
Broadcaster happens with the via the gRPC client you built and error has nothing to do with `H:Err`. Also, the broadcast grpc method is per msg instead of a batch of msgs.